### PR TITLE
feat: TRTLLM FMHAv2 backend for ctx attention

### PR DESCRIPTION
**Pull request by jimmyzho**
## 📌 Description

Porting over the [trtllm fmhav2 library](https://github.com/NVIDIA/TensorRT-LLM/tree/main/cpp/kernels/fmha_v2) to support prefill cases.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [ ] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [ ] I have installed the hooks with `pre-commit install`.
- [ ] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [ ] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **New Features**
  * INT8 quantization and FP8 (E4M3/E5M2) conversion utilities, plus broad packed 8/16‑bit output paths.
  * Hopper GMMA/TMA optimizations and SM90 GMMA/IGMMA helpers for high‑performance kernels.
  * Extensive FMHA v2 tiling/load/store primitives (Q/K/V/O), TMA descriptor management, and paged KV cache.

* **Enhanced Support**
  * Alibi positional-bias params, BF16/mixed-precision conversions, causal/sliding-window masks and multi‑token prediction.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: tests/attention/test_trtllm_prefill_deepseek.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    else:
+        mask = torch.ones(qo_len, kv_len, device=q.device)
+
+    logits = logits.masked_fill(mask.unsqueeze(0).unsqueeze(0) == 0, float("-inf"))
+    # LSE computation: logsumexp over the key dimension (last dim)
+    # logits shape: (batch, num_heads, seq_len, seq_len)
+    lse_ref = torch.logsumexp(logits, -1)  # (batch, num_heads, seq_len)
+    # Transpose to match expected shape (batch, seq_len, num_heads)
+    lse_ref = lse_ref.transpose(1, 2)
+    p = torch.softmax(logits, dim=-1)
```
Comment (weireweire) [inline tests/attention/test_trtllm_prefill_deepseek.py:36]: directly use `lse_ref` to get softmax results may be better?

File: tests/attention/test_trtllm_prefill_deepseek.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        return_lse=True,
+        lse=lse,
+    )
+    # implementation gives [max(s_i), sum(exp(s_i - max(s_i)))], compute lse from this
+    if qkv_dtype == torch.float8_e4m3fn:
+        # For E4M3 the softmax is scaled by 256 (the largest power-of-2 below E4M3_MAX=448.0)
+        descale = 256
+        lse = lse[:, :, :, 0] + torch.log(lse[:, :, :, 1] / descale)
+    else:
+        lse = lse[:, :, :, 0] + torch.log(lse[:, :, :, 1])
```
Comment (weireweire) [inline tests/attention/test_trtllm_prefill_deepseek.py:143]: should this be done by user outside API? and what lse used for outside attention, for training?

Comment (yzh119) [reply tests/attention/test_trtllm_prefill_deepseek.py:143]: I think we should do this inside kernel.

File: csrc/fmha_v2/softmax_bf16.cu [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ * without an express license agreement from NVIDIA CORPORATION or
+ * its affiliates is strictly prohibited.
+ */
+
+#include "softmax_impl.h"
+
+void run_softmax_bf16(void* dst, void const* src, void const* mask, void const* attention_sinks,
+                      void* softmax_sum_d, void* cu_q_seqlens_d, int s_inner, int s_outer, int b,
+                      int h, float softcapping_scale_bmm1, int warps_n, bool has_alibi) {
+  printf("Softmax bf16\n");
```
Comment (yzh119) [inline csrc/fmha_v2/softmax_bf16.cu]: do we expect to keep them?

File: csrc/fmha_v2/softmax_fp8.cu [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+ * its affiliates is strictly prohibited.
+ */
+
+#include "softmax_impl.h"
+
+void run_softmax_e4m3(void* dst, void const* src, void const* mask, void const* attention_sinks,
+                      void* softmax_sum_d, void* cu_q_seqlens_d, int s_inner, int s_outer, int b,
+                      int h, float scale_softmax, float softcapping_scale_bmm1, int warps_n,
+                      bool has_alibi) {
+  printf("Softmax e4m3\n");
```
Comment (yzh119) [inline csrc/fmha_v2/softmax_fp8.cu]: ditto

File: flashinfer/jit/attention/trtllm/generate_kernels.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+        # Sanitize kernel specs
+        specs_expanded = [kspec for kspec in specs_expanded if kspec.sm >= kspec.sm_mma]
+
+        specs_names = [
+            (kspec, *encode_name(kspec))
+            for kspec in specs_expanded
+            # Volta is deprecated in TRT-LLM.
+            if (
+                kspec.sm in [80, 86, 89, 90, 120]
```
Comment (yzh119) [inline flashinfer/jit/attention/trtllm/generate_kernels.py]: We should probably relax it to `kspec.sm >= 80`?

File: flashinfer/jit/attention/fmha_v2/generator_utils.py
Comment (yzh119) [inline flashinfer/jit/attention/fmha_v2/generator_utils.py:1]: Can we rename this file? `setup.py` sounds like a package installation script.

File: flashinfer/jit/attention/fmha_v2/generate_kernels.py
Comment (yzh119) [inline flashinfer/jit/attention/fmha_v2/generate_kernels.py:1]: Also, I encourage renaming `trtllm` module to `fmha_v2` because there are different attention implementation backends in trtllm

File: flashinfer/prefill.py [hunk 6]
```diff
@@ -3551,3 +3552,83 @@ def trtllm_batch_context_with_kv_cache(
         if out_dtype != "nvfp4"
         else FP4Tensor(out, out_scale_factor, o_sf_start_index, query.shape)
     )
+
+
+def trtllm_prefill_deepseek(
```
Comment (yzh119) [inline flashinfer/prefill.py]:
```suggestion
def fmha_v2_prefill_deepseek(
```

File: flashinfer/jit/attention/fmha_v2/generator_utils.py
Comment (jimmyzho) [reply flashinfer/jit/attention/fmha_v2/generator_utils.py:1]: yes, I can rename to something like fmha_v2_generator_utils.py - this was the original generator script in the trtllm repo.

**Review by yzh119 — APPROVED** [review]
LGTM overall, let's merge this one first and implement the in-kernel LSE calculation in a future PR.
